rendezvous/README.md

```diff
@@ -186,7 +186,46 @@ register under a namespace of interest (eg: the relay namespace). This
 can be mitigated by requiring a Proof of Work scheme for client
 registrations.
 
-This is TBD before finalizing the spec.
+If a rendezvous node decides that Proof of Work is required to complete
+a registration, it will send a response with the status code `E_POW_REQUIRED`
+together with a `challenge`.
+
+A client is then expected to follow up with a `ProofOfWork` message, computing
+a hash by concatenating the following:
+
+- The UTF8 bytes of the string `libp2p-rendezvous-pow`
+- The contents of `challenge`
+- The `ns` field
+- The `signedPeerRecord` field
+- The `PeerId` of the rendezvous point we are registering with
+- The `nonce` field
+
+The resulting byte buffer is hashed using SHA256.
+
+We define "difficulty" as the number of zero bytes at the front of the
+hash assuming a big endian encoding. If a rendezvous point considers
+the difficulty too low, it will decline a `ProofOfWork` message with
+a `RegisterResponse` and the status code set to `E_POW_DIFFICULTY_TOO_LOW` together
+with a new `challenge`.
+
+Once the difficulty requirement is met, a rendezvous point will respond with a
+status code of `OK`.
+
+To vary the difficulty, a client can manipulate the `nonce` field.
+It is purposely unspecified, what the required difficulty for a
+rendezvous point is. This allows rendezvous points to vary the difficulty
+based on current load.
+
+Example:
+
+```
+A -> R: REGISTER{my-app, {QmA, AddrA}}
+R -> A: {E_POW_REQUIRED, challenge: 0xDEADBEEFDEADBEEF}
+A -> R: PROOF_OF_WORK{0x00FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF, nonce: 3123452}
+R -> A: {E_DIFFICULTY_TOO_LOW, challenge: 0xBEEFDEADBEEFDEAD}
+A -> R: PROOF_OF_WORK{0x000000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF, nonce: 8572945}
+R -> A: {OK}
```

Comment:
All of this happens in a single substream. This is necessary because the rendezvous point needs to keep context of which challenge they provided to which node. If we were to close the substream after `{E_POW_REQUIRED}` and make the client open a new substream with the PoW, then we wouldn't be able to reliably tell, which challenge we gave them (and we certainly can't trust them to tell us!)